### DISCUSSION TITLE
Access to e.g. getAccountabilityForRole and ItemsService in @directus/extensions-sdk

### DISCUSSION THREAD

**Discussion by danybol** [directus/directus#22782] [Feature Requests]
### Summary

Love Directus, you guys are doing a great job! I just wanted to mention something which would be really helpful for writing extensions. We would like the ability to use the existing getAccountabilityForRole logic in extensions, so that we can override the accountability of the role running a flow (for an 'operation' extension).

In addition, providing access to ItemsService as part of the extensions sdk (rather than having it passed dynamically) would be really useful for writing integration tests for extensions. At the moment we have to mock ItemsService completely.

I see there was a discussion about this a year or so ago here: https://github.com/directus/directus/discussions/5820. I was just wondering if the codebase is now more stable so this could be revisited? I see that getAccountabilityForRole hasn't changed in 7 months now.

### Overriding Role
```
import { getAccountabilityForRole } from '@directus/extensions-sdk'
...
const publicAccountability = await getAccountabilityForRole(null, { database, schema, accountability })
const itemsService = new ItemsService(collection, {
				schema: await getSchema(),
				accountability: publicAccountability
			})
```

### Integration tests

```
import { ItemsService } from '@directus/extensions-sdk'

test("some operation runs against database", () => {
  // some knex setup code
  ...
  const res = await handler({ collection }, { services: { ItemsService } })
  // res contains real items from the database
})
```

### Motivation

Improves testability of extensions, and allows more roles-based use cases such as overriding the role a custom operation is run with (e.g. force a custom extension to run as public role even if is triggered by admin)

### Detailed Design

`import { getAccountabilityForRole, ItemsService } from '@directus/extensions-sdk'`

### Must Have:
- getAccountabilityForRole available from '@directus/extensions-sdk'
- ItemsService available from '@directus/extensions-sdk'

### Drawbacks

Directus internals often change, so would add to maintenance burden.

### Alternatives

Roll-your-own permissions management for each custom extension (which may break if the directus internals break)

### Adoption Strategy

I believe this won't be a breaking change - just moving certain functions into shared code

**Comment by rijkvanzanten** [comment]
It always boils down to the same question of is the additional flexibility in extensions worth the tradeoff of the additional amount of breaking changes we'll have to ship to those same extensions, and are extension builders intending on keeping their extensions up to date if we do 🤔 

For example the `getAccountabilityForRole` util function that we use internally was just changed to support policies (https://github.com/directus/directus/pull/22773/files#diff-324aa8ff71aa96c3c54e181bb23c7a824caac03ed7fc6dbdcab545405bb0c3bb), that means this would have been a breaking change for extensions relying on it, whereas all the things we expose dynamically have had stable APIs.

I think the only solution here is to modularize the API internals a lot more, and make sure that each of those modules has a stable API that we can then safely expose to both the API internals and extensions

**Reply by danybol** [reply]
Thanks for the quick reply, yes as soon as I created the feature request I noticed the new update changes how permissions work! That being said, looking at the diff the interface seems to be the same? For our usecase at least we don't really care about the internals. But in general I definitely see what you mean.

We can work around it for now anyway, but having access to a stable version of some of these internals would make developing extensions much smoother and unlock extra functionality. At the moment I don't think you can create a custom extension which mimics the functionality of the built-in Read Items operation, for example: https://github.com/directus/directus/blob/main/api/src/operations/item-read/index.ts

Thanks again!

**Reply by rijkvanzanten** [reply]
Yeah fully agreed! My current plan is to design a bus between extensions and the host process (eg the API) which can then have a predefined set of operations for the extension to request, that don't have to rely on importing actual functions and internals. That way we can keep being flexible and "break" things in the internals without hindering extensions, as the API surface between host and extension can stay the same. Of course that would have a similar question around what functionality is available through the bus, but those feel easier to answer than trying to directly expose more internal functions

**Reply by falmanna** [reply]
During my work on custom endpoints I realized that it is very hard to construct accountability, specifically setting the permissions of accountability. 

I took a deep dive to the source code and ended up creating this Util. So far it is working well.

```typescript
export async function getUserPermissions(
	context: HookExtensionContext,
	userId: string,
	withAppAccess: boolean = false,
): Promise<Permission[]> {
	const userRepo = await constructRepository(context, "directus_users");
	const user = await userRepo.readOne(userId);
	const roleRepo = await constructRepository(context, "directus_roles");
	const userRole = await roleRepo.readOne(user.role as string);

	const accountability = {
		user: user.id,
		role: user.role as string,
		admin: true,
		app: withAppAccess,
	};

	const permissionsService: PermissionsService = new context.services.PermissionsService({
		schema: await context.getSchema({ database: context.database }),
		knex: context.database,
		accountability: accountability,
	});
	const permissions = (await permissionsService.readByQuery({
		filter: { role: { _eq: accountability.role } },
	})) as Permission[];

	return permissions.map((permission) => {
		const filterContext = {
			$CURRENT_USER: user as any,
			$CURRENT_ROLE: userRole as any,
		};
		permission.permissions = parseFilter(permission.permissions, accountability, filterContext);
		permission.validation = parseFilter(permission.validation, accountability, filterContext);
		permission.presets = parsePreset(permission.presets, accountability, filterContext);

		return permission;
	});
}
```
